span: update constructors to match c++20 draft spec and add lifetimebound attribute

theuni opened the pull request:
Replaces #19382 with a different approach. See [this comment](https://github.com/bitcoin/bitcoin/pull/19382#discussion_r446332852) for the reasoning behind the switch.

-- 

Description from #19382:

See [here](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0936r0.pdf) for more detail on lifetimebound.

This is implemented using preprocesor macros rather than configure checks in order to keep span.h self-contained.

The ```[[clang::lifetimebound]]``` syntax was chosen over ```__attribute__((lifetimebound))``` because the former is more flexible and works to guard ```this``` as well as function parameters, and also because at least for now, it's available only in clang.

There are currently no violations in our codebase, but this can easily be tested by inserting one like this somewhere and compiling with a modern clang:
```c++
Span<const int> bad(std::vector<int>{1,2,3});
```

The result:
> warning: temporary whose address is used as value of local variable 'bad' will be destroyed at the end of the full-expression [-Wdangling]
    Span<const int> bad(std::vector<int>{1,2,3});
```

theuni committed:
span: add lifetimebound attribute

See http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0936r0.pdf for
reference.

This helps to guard against dangling references caused by construction from
temporaries such as:

    Span<const int> sp(std::vector<int>{1,2,3});

theuni committed:
span: (almost) match std::span's constructor behavior

c++20's draft of std::span no longer includes move constructors.

sipa commented on src/span.h:
```diff
@@ -78,8 +96,19 @@ class Span
+    constexpr Span(const V& other SPAN_ATTR_LIFETIMEBOUND,
+        typename std::enable_if<!is_Span<V>::value &&
+                                std::is_convertible<typename std::remove_pointer<decltype(std::declval<const V&>().data())>::type (*)[], C (*)[]>::value &&
+                                std::is_convertible<decltype(std::declval<V&>().size()), std::size_t>::value, std::nullptr_t>::type = nullptr)
```
Nit: `declval<const V&>`. Not that I expect size() to differ between const and non-const objects...

theuni replied:
Fixed, thanks.

sipa commented:
ACK apart from nit above.

I've tested this by rebasing #13062 and #19326 on top, and compiling with Clang 10; no warnings. Adding an obvious `Span<const int> bad{std::vector<int>{}};` does produce a warning.

It does seem that no warning is produced when `MakeSpan` is used in between, and I can't get that fixed. I think that's fine, but it'd be nice to improve upon.

practicalswift commented:
Concept ACK: `[[clang::lifetimebound]]` is great! :)

I think this will be useful also outside of `src/span.{cpp,h}`. What about adding it to `src/attributes.h` as `LIFETIMEBOUND`?

```patch
diff --git a/src/attributes.h b/src/attributes.h
index 45099bd8b..9d5c1d44a 100644
--- a/src/attributes.h
+++ b/src/attributes.h
@@ -19,4 +19,14 @@
 #  endif
 #endif
 
+#if defined(__clang__)
+#  if __has_attribute(lifetimebound)
+#    define LIFETIMEBOUND [[clang::lifetimebound]]
+#  else
+#    define LIFETIMEBOUND
+#  endif
+#else
+#  define LIFETIMEBOUND
+#endif
+
 #endif // BITCOIN_ATTRIBUTES_H
```

jonatack commented:
Concept ACK

theuni commented:
Updated and squashed.

@practicalswift I'd rather wait to add it there until there's a use for it, if that's ok. By the time we have it to use somewhere else, it may not be clang-only anymore.

I did change it to use ```__has_cpp_attribute``` to match our existing attribute check, though.

theuni commented:
Sorry for the pushing/rebasing noise, done now.

> It does seem that no warning is produced when MakeSpan is used in between, and I can't get that fixed. I think that's fine, but it'd be nice to improve upon.

@sipa It turns out that this warning _does_ work for MakeSpan, but only in c++17 mode and above:

```
$ clang++ spantest.cpp -o spantest -std=c++17
```
> spantest.cpp:10:26: warning: temporary whose address is used as value of local variable 'temp' will be destroyed at the  end of the full-expression [-Wdangling]
>    auto temp = MakeSpan(std::vector<int>{1,2,3});

 I've gone ahead and added the annotations there as well, so they should just magically work when we switch to c++17.

sipa commented:
> so they should just magically work when we switch to c++17.

I like magic.

ACK 1d58cc7cb040a70f768b632f294db4e0797d3a34

MarcoFalke commented:
Instead of magic, I'd rather remove them and add the corresponding constructors from c++20 when we switched to c++17

second-commit-only review ACK 1d58cc7cb040a70f768b632f294db4e0797d3a34

MarcoFalke commented on src/span.h:
```diff
@@ -18,6 +18,16 @@
 
+#if defined(__clang__)
+#if __has_attribute(lifetimebound)
+#define SPAN_ATTR_LIFETIMEBOUND [[clang::lifetimebound]]
```
```suggestion
#define ATTR_LIFETIMEBOUND [[clang::lifetimebound]]
```

Can this be named a bit more generic, so that it can be used in other places such as #19426 ?

theuni replied:
I namespaced it to avoid colliding with a more general define. As ```span.h``` is designed to be an abstraction, I'd really prefer to keep it an island and not require that it include another header.

Since @sipa was quick to point out a few places where this could already be useful, how about going with @practicalswift's [idea](https://github.com/bitcoin/bitcoin/pull/19387#issuecomment-650752959) of adding a second, more generic define in ```attributes.h``` ?

sipa replied:
@theuni Not sure how to weigh code duplication vs. a dependency on attributes.h. I'm fine with either.

sipa commented:
@MarcoFalke Well, no reason why we can't have both. The attribute as argument to `MakeSpan` makes sense - even if the compiler doesn't do anything useful with it - as it documents behavior to the user as well. And perhaps clang 123091137.4 will support it even in C++11 mode - who knows.

sipa commented:
There are a lot more cases in the codebase where a data type stores a reference or pointer to another object provided to the constructor exist. For example:

* In `src/stream.h`: `OverrideStream`, `CVectorReader`, `BitStreamReader`, `BitStreamWriter`
* In `src/script/sign.h`: `MutableTransactionSignatureCreator`
* In `src/serialize.h`: `Wrapper`, `Using`
* In `src/wallet/rpcwallet.cpp`: `DescribeWalletAddressVisitor`
* In `src/key_io.cpp`: `DestinationEncoder`
* In `src/dbwrapper.h`: `CDBBatch`, `CDBIterator`
* In `src/flatfile.h`: `FlatFileSeq`
* In `src/hash.h`: `CHashVerifier`
* In `src/miner.h`: `CBlockAssembler`
* In `src/net_processing.h`: `PeerLogicValidation`
* In `src/scheduler.h`: `SingleThreadedSchedulerClient`
* In `src/sync.h`: `CSemaphoreGrant`
* In `src/validation.h`: `CScriptCheck`, `ChainstateManager`,  `CChainState`

Several of these accept const lvalue references, and store them, and are at risk already. Others take in a mutable lvalue reference or pointers, but would make sense to support rvalue/universal references as input (so that temporaries can be passed to it). In all those cases, having a `lifetimebound` attribute would be useful.

So I think that means we should just put it in attributes.h. There is plenty of potential for it.

jonatack reviewed (commented):
Non-expert code review ACK 1d58cc7cb040a

Looked at the linked doc and other docs/implementations I could find, abseil/span.h etc. This compiled fine including with it added to `src/attributes.h`. Unfortunately I am currently stuck on Debian Clang 6 which is less helpful.

jonatack commented:
Upgraded to Clang 11 and now see the -Wdangling warning mentioned in the PR description.
```
wallet/rpcwallet.cpp:42:22: warning: temporary whose address is used as value of local variable
'a_bad_vec' will be destroyed at the end of the full-expression [-Wdangling]
Span<const int> a_bad_vec(std::vector<int>{1,2,3});
                     ^~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

theuni committed:
Add lifetimebound to attributes for general-purpose usage

Co-authored-by: practicalswift <practicalswift@users.noreply.github.com>

sipa commented:
ACK e3e7446305329ce96e9cf5f5161658eb2e1ea888

sipa commented:
Sorry, I misclicked.

jonatack commented on src/attributes.h:
```diff
@@ -19,4 +19,14 @@
+#  endif
+#else
+#  define LIFETIMEBOUND
+#endif
```
Build warning (in clang 11.0, but not gcc 9.3)
```
In file included from ./util/system.h:17:
./attributes.h:30:7: warning: extra tokens at end of #endif directive [-Wextra-tokens]
#endif#if defined(__clang__)
      ^
      //
1 warning generated.
```
```
((HEAD detached at origin/pr/19387))$ clang --version
Debian clang version 11.0.0-++20200708111116+1f84ace3c72-1~exp1~20200708091733.3345 
Target: x86_64-pc-linux-gnu
```

theuni replied:
I don't understand what this is complaining about. Looks like some subtle whitespace thing, but it's all fine as far as I can tell. Why wouldn't it warn about the block above as well?

jonatack replied:
Same. Maybe that version of clang is buggy if no one else got it.

theuni replied:
I can't hit this with clang 10.

Does it spew a million warnings for our includes? That would make sense, at least. But if this is the only one there must be a reason.

If this is a real clang bug, free to bisect and report upstream since you're building from git, so we don't hit this when 11 is released :)

jonatack replied:
Installed the new clang nightly updates to this version, built, and the warning is gone. (Yay!)
```
$ clang --version
Debian clang version 11.0.0-++20200709111134+dc4a6f5db4f-1~exp1~20200709091753.3347

$ uname -a
Linux 5.7.0-1-amd64 #1 SMP Debian 5.7.6-1 (2020-06-24) x86_64 GNU/Linux
```

jonatack reviewed (commented):
ACK e3e7446 change since last review is adding `[[clang::lifetimebound]]` as `LIFETIMEBOUND` to src/attributes.h as suggested in https://github.com/bitcoin/bitcoin/pull/19387#issuecomment-650752959.

sipa commented:
Mostly a reminder for myself: `WithParams`, `ParamsStream`, and `ParamsWrapper` in https://github.com/bitcoin/bitcoin/pull/19503 could also use `LIFETIMEBOUND`.

sipa commented:
Anyone else feel like reviewing this?

JeremyRubin commented:
I'm curious how this plays with c++17, if/when we switch to std::span will we be able to keep the annotation somehow?

sipa commented:
@JeremyRubin We can keep annotations on our own functions, but at least MakeSpan won't be needed anymore. On `std::span` we can't put our own annotations of course, but presumably the stdlib that ships with c++20 supporting clang will have this annotation built-in for its span type.

JeremyRubin commented:
ok can revisit in 10 years ;)

ajtowns commented:
ACK e3e7446305329ce96e9cf5f5161658eb2e1ea888 (drive by; only a quick skim of code and some basic sanity checks)

MarcoFalke commented:
review ACK e3e7446305329ce96e9cf5f5161658eb2e1ea888 🔗

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK e3e7446305329ce96e9cf5f5161658eb2e1ea888 🔗
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi3Mgv+Lgqd8NfB61HB1uzfd9S9SLQ0uPmWfMMNwBxuueD7YyuhvsNS2sto+2cK
3dwf5m0EkQHymOEWtPZRAR1TFgr+phnM7ky4/MEJkECSiNWWHWCNqFybzpNvoPjt
lPAvduAM9OO4ZiHRPgqTsVGhz01hGpN7D+Hjo7bMu5y5yv5UFaOhz90earJ/44AI
cWiZTw2YVmO9MZBWnBUUFUwj7OwBN5kn8SXHhe6/EoZkECpD/aIMXElUB5gmu+l+
ko11vuJdzl4dQpf9v6oPl/0kSACwwf8nEpAM+LjuHOMcGZKrCmwPwmRF8tRmLAY5
p/BwPR3wvA7EQaqJOVnZ+S1GcAEpFt5J8eQ3Vn0Z0qyG+7LQ6om3TSkWEP2fqZQ7
h0L/ZJwbSd5nuvVV4D3du+P0SoKW6nAN1W8Iz+Yrt45CnoZ1n2v3dVLUw0S/HyIF
BIAVD9e2T6wznmOsOsZJpX38l4fORnwXpY7jH2mO7nFJmOQ8r8dRiVyR/obkk/Sq
psWzYP6c
=x28h
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c85bdf628bd486c6b0345f5d023467302549bb7e57ac1eb39166395d465fbb11  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c85bdf628bd486c6b0345f5d023467302549bb7e57ac1eb39166395d465fbb11f010759ff274f35455cc1b7856856166e1fd08fff01050e44f425b7710e62cfad30f1c8fd03008f1045fbe6796f008ad764d5210e325220083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b799ee64b5c5623a0ec30b4cacb6c46d08f1045fbe679af00857b133903b9e12a50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010e8fc12e4001fc3e9417be576a2e760bb08f1045fbe6796f0085d7f19ef8950dda60083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f225265eb9e58810a7be98668d99aa8108f1045fbe6796f008a982b291931fa1af0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>
